server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/StoreMailHelper.java

```diff
@@ -0,0 +1,98 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.queue.rabbitmq.view.cassandra;
+
+import java.time.Clock;
+import java.time.Instant;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.inject.Inject;
+
+import org.apache.james.queue.rabbitmq.MailQueueName;
+import org.apache.james.queue.rabbitmq.view.cassandra.model.BucketedSlices.BucketId;
+import org.apache.james.queue.rabbitmq.view.cassandra.model.EnqueuedMail;
+import org.apache.james.queue.rabbitmq.view.cassandra.model.MailKey;
+import org.apache.mailet.Mail;
+
+class StoreMailHelper {
+
+    private final EnqueuedMailsDAO enqueuedMailsDao;
+    private final BrowseStartDAO browseStartDao;
+    private final CassandraMailQueueViewConfiguration configuration;
+    private final Clock clock;
+    private final Set<MailQueueName> initialInserted;
```

Comment:
what is that? a cache?

Reply:
The idea is to avoid the expensive `INSERT IF NOT EXISTS` operation on each write.

`INSERT IF NOT EXISTS` is an expensive operation for several reasons:
 - Performs read before writes, which is a Cassandra anti-pattern
 - Triggers PAXOS consensus between nodes
 - ... all of that can induce a 1000x perf penalty.

Lightweight conditions have a cost.

The good news for us is that we only need to try to insert the timestamp on the first enqueued email (subsequent emails will be after that browse start pointer and will not need to initialize it).

Thus each James node can keep track of whether it already initialized the browse start on that queue, and pay only once that cost.

Reply:
First make it correct, then make it concise, then make it fast.
Please drop this cache for now.